analyzer/src/main/kotlin/managers/Pnpm.kt

```diff
@@ -0,0 +1,101 @@
+import org.ossreviewtoolkit.analyzer.managers.utils.mapDefinitionFilesForPnpm
+import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
+import org.ossreviewtoolkit.model.config.RepositoryConfiguration
+import org.ossreviewtoolkit.model.jsonMapper
+import org.ossreviewtoolkit.utils.common.Os
+import org.ossreviewtoolkit.utils.common.realFile
+
+/**
+ * The [fast, disk space efficient package manager](https://pnpm.io/).
+ */
+class Pnpm(
+    name: String,
+    analysisRoot: File,
+    analyzerConfig: AnalyzerConfiguration,
+    repoConfig: RepositoryConfiguration
+) : Npm(name, analysisRoot, analyzerConfig, repoConfig) {
+    class Factory : AbstractPackageManagerFactory<Pnpm>("PNPM") {
+        override val globsForDefinitionFiles = listOf("package.json", "pnpm-lock.yaml")
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Pnpm(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    /**
+     * PNPM symlinks workspace modules in the `node_modules` directory, which will result in cyclic symlinks of
+     * `node_module` directories. Limit the search depth to '2' in this case as all packages are in a direct
+     * subdirectory of the `node_modules` directory, thanks to the `--shamefully-hoist` install option.
+     */
+    override val modulesSearchDepth = 2
+
+    override fun hasLockFile(projectDir: File) = hasPnpmLockFile(projectDir)
+
+    override fun File.isWorkspaceDir() = realFile() in findWorkspaceSubmodules(analysisRoot)
+
+    override fun command(workingDir: File?) = if (Os.isWindows) "pnpm.cmd" else "pnpm"
+
+    override fun getVersionRequirement(): Requirement = Requirement.buildNPM("5.* - 7.8.*")
+
+    override fun mapDefinitionFiles(definitionFiles: List<File>) = mapDefinitionFilesForPnpm(definitionFiles).toList()
+
+    override fun runInstall(workingDir: File) =
+        run(
+            "install",
+            "--ignore-pnpmfile",
+            "--ignore-scripts",
+            "--frozen-lockfile", // Use the existing lockfile instead of updating an outdated one.
+            "--shamefully-hoist", // Build a similar node_modules structure as NPM and Yarn does.
+            workingDir = workingDir
+        )
+
+    override fun beforeResolution(definitionFiles: List<File>) =
+        // We do not actually depend on any features specific to a PNPM version, but we still want to stick to a
+        // fixed minor version to be sure to get consistent results.
+        checkVersion()
+
+    override fun getRemotePackageDetails(workingDir: File, packageName: String): JsonNode {
+        val process = run(workingDir, "view", "--json", packageName)
+
+        return jsonMapper.readTree(process.stdout)
+    }
+
+    override fun findWorkspaceSubmodules(moduleDir: File): List<File> =
+        submodulesCache.getOrPut(moduleDir.absolutePath) {
```

Comment:
Ideally, the subclass does not need to know about caching. This can be achieved as follows:
The `findWorkspaceSubmodules()` function in `Npm` is made non-open again. It checks whether the requested path is already in the cache. If not, it delegates to a new function `loadWorkspaceSubmodules()` (or whatever name fits) that does the actual retrieving of modules. The result is then stored in the cache and returned.

`Pnpm` only overrides this `load` function. The map with the cache can then be private in `Npm`. Then only care has to be taken that `Pnpm` never invokes the `load` function itself, but accesses all submodules via the `findWorkspaceSubmodules()` function.